lib/shoulda/matchers/action_controller/permit_matcher.rb

```diff
@@ -197,32 +213,49 @@ def matches?(controller)
         end
 
         def failure_message
-          "Expected controller to permit #{unpermitted_params.to_sentence}, but it did not."
+          "Expected #{verb.upcase} ##{action} to #{expectation},\nbut #{reality}."
         end
         alias failure_message_for_should failure_message
 
         def failure_message_when_negated
-          "Expected controller not to permit #{verified_permitted_params.to_sentence}, but it did."
+          "Expected #{verb.upcase} ##{action} not to #{expectation},\nbut it did."
         end
         alias failure_message_for_should_not failure_message_when_negated
 
         protected
 
-        attr_reader :controller, :double_collection, :action, :verb,
-          :request_params, :expected_permitted_params, :context
+        attr_reader :controller, :double_collections_by_param, :action, :verb,
+          :request_params, :expected_permitted_params, :context, :subparameter,
+          :parameters_doubles
+
+        def expectation
+          message = "restrict parameters "
+
+          if subparameter
+            message << " for #{subparameter.inspect}"
+          end
+
+          message << "to " + format_param_names(expected_permitted_params)
 
-        def set_double_collection
-          @double_collection =
-            Doublespeak.double_collection_for(::ActionController::Parameters)
+          message
+        end
 
-          @double_collection.register_stub(:require).to_return(&:object)
-          @double_collection.register_proxy(:permit)
+        def reality
+          if actual_permitted_params.empty?
+            "it did not restrict any parameters"
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.